### PR TITLE
PrairieView Acq added

### PR DESCRIPTION
**Pull request by kushalbakshi**
Once changes here look good, I can add a reference to it under requirements for `workflow-calcium-imaging`

### CODE REVIEW

File: element_calcium_imaging/scan.py [hunk 1]
```diff
@@ -432,6 +441,47 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
...
+                    dict(
+                        key,
+                        field_idx = plane_idx,
+                        px_height = scan["height_in_pixels"],
+                        px_width = scan["width_in_pixels"],
+                        um_height = scan["height_in_um"],
+                        um_width = scan["width_in_um"],
+                    )
+                    for plane_idx in range(scan["num_planes"])
+                ]
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/scan.py:483]:
```suggestion
                dict(
                    key,
                    field_idx = plane_idx,
                    px_height = scan["height_in_pixels"],
                    px_width = scan["width_in_pixels"],
                    um_height = scan["height_in_um"],
                    um_width = scan["width_in_um"],
                    )
                    for plane_idx in range(scan["num_planes"])
```

File: element_calcium_imaging/scan.py [hunk 2]
```diff
@@ -432,6 +441,47 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
...
                     ]
                 )
+        elif acq_software == "PrairieView":
+            import prairieviewreader
+            
+            # Read the scan
+            scan_filepaths = get_prairieview_files(key)
+            scan = prairieview.get_pv_metadata(scan_filepaths)
+
+            # Insert in ScanInfo
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/scan.py:451]: This comment is not informative
```suggestion
```

File: element_calcium_imaging/scan.py [hunk 3]
```diff
@@ -180,6 +188,7 @@ class ScanInfo(dj.Imported):
     z=null               : float     # (um) ScanImage's 0 point in the motor coordinate system
     fps                  : float     # (Hz) frames per second - Volumetric Scan Rate 
     bidirectional        : boolean   # true = bidirectional scanning
+    bidirectional_z=null : booleaen  # true = bidirectional z-stack for PrairieView
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/scan.py:191]: Hm... This worked? So probably did not get tested.
```suggestion
    bidirectional_z=null : boolean  # true = bidirectional z-stack for PrairieView
```

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

File: element_calcium_imaging/scan.py [hunk 4]
```diff
@@ -432,6 +441,44 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
                         for plane_idx in range(nd2_file.sizes.get("Z", 1))
                     ]
                 )
+        elif acq_software == "PrairieView":
+            import prairieviewreader
```
Comment (kabilar) [inline element_calcium_imaging/scan.py:445]: I am out of the loop.  Is this a standard python package?  I don't see it on PyPI or GitHub.

**Comment by kabilar** [comment]
Please update version and changelog.

File: element_calcium_imaging/scan.py (hunk 4 above)
Comment (ttngu207) [reply element_calcium_imaging/scan.py:445]: I'm wondering the same thing, @kushalbakshi can you clarify

File: element_calcium_imaging/scan.py [hunk 5]
```diff
@@ -180,6 +188,7 @@ class ScanInfo(dj.Imported):
     z=null               : float     # (um) ScanImage's 0 point in the motor coordinate system
     fps                  : float     # (Hz) frames per second - Volumetric Scan Rate 
     bidirectional        : boolean   # true = bidirectional scanning
+    bidirectional_z=null : boolean  # true = bidirectional z-stack for PrairieView
```
Comment (ttngu207) [inline element_calcium_imaging/scan.py:191]:
```suggestion
    bidirectional_z=null : boolean   # true = bidirectional in z-direction
```

File: element_calcium_imaging/scan.py (hunk 4 above)
Comment (kabilar) [reply element_calcium_imaging/scan.py:445]: Ah I now see it here: https://github.com/kushalbakshi/prairieview-reader

Comment (kabilar) [reply element_calcium_imaging/scan.py:445]: Perhaps we should add this function to [element-interface](https://github.com/datajoint/element-interface) or `element-calcium-imaging`.

Comment (kushalbakshi) [reply element_calcium_imaging/scan.py:445]: Yeah I can add it to `element-interface` and make a PR there if you prefer

Comment (kabilar) [reply element_calcium_imaging/scan.py:445]: That would be great.  Thanks!

**Review by ttngu207 — APPROVED** [review]

File: element_calcium_imaging/scan.py [hunk 6]
```diff
@@ -443,41 +444,40 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
...
                 dict(
                     key,
-                    field_idx = plane_idx,
-                    px_height = pvscan_info["height_in_pixels"],
-                    px_width = pvscan_info["width_in_pixels"],
-                    um_height = pvscan_info["height_in_um"],
-                    um_width = pvscan_info["width_in_um"],
-                    )
-                    for plane_idx in range(pvscan_info["num_planes"])
+                    field_idx=plane_idx,
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/scan.py:471]: This seems a little strange that the same pvscan information is populated for all planes

File: element_calcium_imaging/scan.py [hunk 7]
```diff
@@ -443,41 +444,40 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
...
-                    nframes = pvscan_info["num_frames"],
-                    rois = pvscan_info["num_rois"],
-                    x = pvscan_info["x_pos"],
-                    y = pvscan_info["y_pos"],
-                    z = pvscan_info["z_pos"],
-                    fps = pvscan_info["frame_rate"],
-                    usecs_per_line = pvscan_info["usecs_per_line"],
-                    scan_datetime = pvscan_info["scan_datetime"],
-                    scan_duration = pvscan_info["scan_duration"],
+                    nfields=pvscan_info["num_fields"],
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/scan.py:453]: Isn't `nfields` already in `ScanInfo`?

File: element_calcium_imaging/scan.py (hunk 6 above)
Comment (kushalbakshi) [reply element_calcium_imaging/scan.py:471]: The x and y coordinates do not seem to change over time. This system is only acquiring in different z-planes over time.

File: element_calcium_imaging/scan.py (hunk 7 above)
Comment (kushalbakshi) [reply element_calcium_imaging/scan.py:453]: We're populating the `nfields` in `ScanInfo` here

Comment (dimitri-yatsenko) [reply element_calcium_imaging/scan.py:453]: oops. Got it.

**Review by kabilar — CHANGES_REQUESTED** [review]
Thanks @kushalbakshi!  Great work.  Minor comment below.  Please test an example ingestion prior to merging.  Let's hold off on merging this pull request until [element-interface #44](https://github.com/datajoint/element-interface/pull/44) is merged.

**Comment by kushalbakshi** [comment]
@kabilar @dimitri-yatsenko I've updated `scan.py` to prevent any errors while populating the `ScanInfo` table.

File: element_calcium_imaging/scan.py (hunk 6 above)
Comment (kabilar) [reply element_calcium_imaging/scan.py:478]: @kushalbakshi Perhaps the example file does not represent all cases.  Is it possible to extract the x position, y position, and dimensions of the fields at each depth?

**Review by kabilar — APPROVED** [review]
Thanks @kushalbakshi.  Great work!  Holding off on merging until Dimitri's point above is resolved.

File: element_calcium_imaging/scan.py (hunk 6 above)
Comment (kushalbakshi) [reply element_calcium_imaging/scan.py:478]: The only x and y coordinates present in the metadata file are for the initial position when the scan is started and these coordinates are the center of the scanning field. @kabilar and I came to the conclusion that this scanner likely scans the same x and y coordinates for the entire scan. In the z-plane, it scans through a user-defined number of planes (ndepths). Hopefully this clarifies it a bit further, @dimitri-yatsenko

Comment (kabilar) [reply element_calcium_imaging/scan.py:478]: Thanks for checking @kushalbakshi.  I will go ahead and merge.

**Pull request merged**